src/wp-includes/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -371,10 +379,51 @@ public function register_script_modules() {
 	 * Adds the necessary hooks for the Interactivity API.
 	 *
 	 * @since 6.5.0
+	 * @since 6.9.0 Adds support for client-side navigation in script modules.
 	 */
 	public function add_hooks() {
 		add_filter( 'script_module_data_@wordpress/interactivity', array( $this, 'filter_script_module_interactivity_data' ) );
 		add_filter( 'script_module_data_@wordpress/interactivity-router', array( $this, 'filter_script_module_interactivity_router_data' ) );
+		add_filter( 'wp_script_attributes', array( $this, 'add_load_on_client_navigation_attribute_to_script_modules' ), 10, 1 );
+	}
+
+	/**
+	 * Adds the `data-wp-router-options` attribute to script modules that
+	 * support client-side navigation.
+	 *
+	 * This method filters the script attributes to include loading instructions
+	 * for the Interactivity API router, indicating which modules can be loaded
+	 * during client-side navigation.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param array $attributes The script tag attributes.
+	 * @return array The modified script tag attributes.
+	 */
+	public function add_load_on_client_navigation_attribute_to_script_modules( $attributes ) {
+		if (
+			isset( $attributes['type'] ) && 'module' === $attributes['type'] && isset( $attributes['id'] )
+			&& array_key_exists( $attributes['id'], $this->script_modules_that_can_load_on_client_navigation )
+		) {
+			$attributes['data-wp-router-options'] = wp_json_encode( array( 'loadOnClientNavigation' => true ) );
+		}
+		return $attributes;
+	}
+
+	/**
+	 * Marks a script module as compatible with client-side navigation.
+	 *
+	 * This method registers a script module to be loaded during client-side
+	 * navigation in the Interactivity API router. Script modules marked with
+	 * this method will have the `loadOnClientNavigation` option enabled in the
+	 * `data-wp-router-options` directive.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param string $script_module_id The script module identifier.
+	 */
+	public function add_client_navigation_support_to_script_module( $script_module_id ) {
+		$this->script_modules_that_can_load_on_client_navigation[ $script_module_id . '-js-module' ] = true;
```

Comment:
Let's have this be exclusively about the Script Module ID, and leave all business related to the HTML to the filter.

```suggestion
		$this->script_modules_that_can_load_on_client_navigation[ $script_module_id ] = true;
```

Reply:
Done in https://github.com/WordPress/wordpress-develop/pull/10357/commits/3694f168bbeb8a8e3ed225a3fe98b80d7af175d5.